ktor-server/jvm/src/io/ktor/server/application/plugins/api/ServerPlugin.kt

```diff
@@ -12,31 +12,56 @@ import io.ktor.server.request.*
 import io.ktor.server.response.*
 import io.ktor.util.*
 import io.ktor.util.pipeline.*
-import kotlin.random.Random
+import kotlin.random.*
 
 /**
  * A plugin that embeds into the HTTP pipeline and extends Ktor functionality.
  **/
-public abstract class ServerPlugin<Configuration : Any> private constructor(
-    pluginFactory: ServerPluginFactory<Configuration>
+public sealed class ServerPlugin<Configuration : Any> private constructor(
+    internal val key: AttributeKey<out ServerPlugin<Configuration>>
 ) : PluginContext {
+
+    public abstract class ApplicationPlugin<Configuration : Any>(
+        pluginFactory: ServerApplicationPluginFactory<Configuration>
+    ) : ServerPlugin<Configuration>(pluginFactory.key) {
+
+        /**
+         * A configuration for the current plugin. It may hold required data and any functionality that can be used by your
+         * plugin.
+         **/
+        public abstract val pluginConfig: Configuration
+    }
+
+    public abstract class RoutingScopedPlugin<Configuration : Any>(
+        pluginFactory: ServerRoutingScopedPluginFactory<Configuration>
+    ) : ServerPlugin<Configuration>(pluginFactory.key) {
+
+        /**
+         * A configuration for the current plugin for this pipeline. It may hold required data and any
+         * functionality that can be used by your plugin.
+         **/
+        public val PipelineContext<*, ApplicationCall>.pluginConfig: Configuration
+            get() = call.attributes[configKey]
+
+        /**
+         * A configuration for the current plugin for this pipeline. It may hold required data and any
+         * functionality that can be used by your plugin.
+         **/
+        public val CallHandlingContext.pluginConfig: Configuration
+            get() = context.pluginConfig
```

Comment:
Why these extensions are declared inside `RoutingScopedPlugin`?
If they are useful maybe it's worth making them available for any plugin?